### PR TITLE
Implement Gradient Clipping, LR Scheduling, and Structural Metrics

### PR DESCRIPTION
**Pull request by pimenol**
-  Integrated gradient clipping and robust NaN/Inf gradient checks into the core TTT loop to prevent model corruption during adaptation.

- Added support for Learning Rate Scheduling (Cosine Annealing and Cosine with Warmup) in TTTConfig.

- Added a new configuration (GRAD_CLIP_ESMFOLD_TTT_CFG)

- Implement scaled_pseudo_perplexity

- Enhanced _ttt_eval_step to calculate TM-score and LDDT metrics in real-time if a reference structure is provided.

- Optimized evaluation performance by implementing reusable temporary file handling.

- Updated read_msa and process_msa_seq with an option to handle lowercase characters (insertions) for better compatibility with different MSA formats.

- Implemented a comprehensive test suite in tests/test_esmfold_ttt.py and a runner script tests/run_test.sh to ensure the reliability of the TTT pipeline and metric calculation logic.

- Refactored calculate_tm_score to support both TMscore and TMalign executables.

### CODE REVIEW

File: proteinttt/models/esmfold.py [hunk 1]
```diff
@@ -72,36 +85,47 @@ def _ttt_eval_step(
...
+        if correct_pdb_path is not None:
+            if self._ttt_temp_pdb_path is None:
+                with tempfile.NamedTemporaryFile(mode='w', delete=False, suffix='.pdb') as tmp_file:
+                    self._ttt_temp_pdb_path = Path(tmp_file.name)
+            
+            pdb_str_to_write = pdb_str[0] if isinstance(pdb_str, list) else pdb_str
+            with open(self._ttt_temp_pdb_path, 'w', buffering=8192) as f:
+                f.write(pdb_str_to_write)
+            
+            tm_score = calculate_tm_score(self._ttt_temp_pdb_path, correct_pdb_path)
```
Comment (anton-bushuiev) [inline proteinttt/models/esmfold.py]: Please calculate tm_score only if self.ttt_cfg.tmalign_path is not None:

File: proteinttt/base.py [hunk 2]
```diff
@@ -430,6 +463,7 @@ def ttt(
                     ttt_step_time=ttt_step_time,
                     score_seq_time=score_seq_time,
                     eval_step_time=eval_step_time,
+                    lr=optimizer.param_groups[0]["lr"],
```
Comment (anton-bushuiev) [inline proteinttt/base.py]: Maybe we can only store lr if the schedule is used. Otherwise it is constant and we don't have to store it

**Review by anton-bushuiev — CHANGES_REQUESTED** [review]

File: proteinttt/base.py [hunk 3]
```diff
@@ -742,16 +799,14 @@ def _ttt_get_state(self) -> T.Any:
...
         modules such as in the case of LoRA.
 
-        TODO: Optimize memory by only saving modules from _ttt_get_trainable_modules()
-
         Returns:
             Dictionary mapping module names to their copied states
         """
         state = {}
-
+        trainable_modules = set(self._ttt_get_trainable_modules())
```
[the remark this answers is not in the document]
Comment (anton-bushuiev) [reply proteinttt/base.py]: Ideally using a different sequence for step (2)
